pkg/ccl/changefeedccl/changefeed_test.go

```diff
@@ -5612,6 +5632,68 @@ func TestChangefeedCaseInsensitiveOpts(t *testing.T) {
 	t.Run(`sinkless`, sinklessTest(testFn))
 }
 
+func TestChangefeedEndTime(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	testFn := func(t *testing.T, db *gosql.DB, f cdctest.TestFeedFactory) {
+		sqlDB := sqlutils.MakeSQLRunner(db)
+
+		sqlDB.Exec(t, "CREATE TABLE foo (a INT PRIMARY KEY)")
+		sqlDB.Exec(t, "INSERT INTO foo VALUES (1), (2), (3)")
+
+		var tsAfterInitialInsert string
+		sqlDB.QueryRow(t, "SELECT cluster_logical_timestamp()").Scan(&tsAfterInitialInsert)
+
+		feed := feed(t, f, "CREATE CHANGEFEED FOR foo WITH end_time = $1", tsAfterInitialInsert)
+
+		defer closeFeed(t, feed)
+		jobFeed := feed.(cdctest.EnterpriseTestFeed)
+		require.NoError(t, jobFeed.WaitForStatus(func(s jobs.Status) bool {
+			return s == jobs.StatusSucceeded
+		}))
+
+		assertPayloads(t, feed, []string{
+			`foo: [1]->{"after": {"a": 1}}`,
+			`foo: [2]->{"after": {"a": 2}}`,
+			`foo: [3]->{"after": {"a": 3}}`,
+		})
+	}
+	t.Run(`enterprise`, enterpriseTest(testFn))
+	t.Run(`cloudstorage`, cloudStorageTest(testFn))
+	t.Run(`kafka`, kafkaTest(testFn))
+	t.Run(`webhook`, webhookTest(testFn))
+}
+
+func TestChangefeedOnlyInitialScan(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	testFn := func(t *testing.T, db *gosql.DB, f cdctest.TestFeedFactory) {
+		sqlDB := sqlutils.MakeSQLRunner(db)
+
+		sqlDB.Exec(t, "CREATE TABLE foo (a INT PRIMARY KEY)")
+		sqlDB.Exec(t, "INSERT INTO foo VALUES (1), (2), (3)")
+
+		feed := feed(t, f, "CREATE CHANGEFEED FOR foo WITH initial_scan_only")
```

Comment:
In both of these tests it'd be nice to have an extra `INSERT` statement right after this `CREATE CHANGEFEED` / end time and then ensure that no kv events other than the initial scan events were emitted.  

`assertPayloads` alone wouldn't be able to assert "nothing extra was sent" though, one option would be a separate goroutine that consumes from the sink ([ex](https://github.com/cockroachdb/cockroach/blob/1eafcc40c352ae9a4ffb9bddf3999a9cf002a4ad/pkg/ccl/changefeedccl/changefeed_test.go#L1219)) and increments a counter for the number of events its seen and ensuring that that number didn't exceed 3.

Reply:
Thanks for this suggestion! I was actually looking to carry out a test similar to the one you described above, but I wasn't sure if it was possible. I will look into improving these tests though 💯